runtime/src/snapshot_utils.rs

```diff
@@ -996,6 +1540,75 @@ pub fn bank_to_snapshot_archive<P: AsRef<Path>, Q: AsRef<Path>>(
     Ok(package.tar_output_file)
 }
 
+/// Convenience function to create an incremental snapshot archive out of any Bank, regardless of
+/// state.  The Bank will be frozen during the process.
+///
+/// Requires:
+///     - `bank` is complete
+///     - `bank`'s slot is greater than `full_snapshot_slot`
+pub fn bank_to_incremental_snapshot_archive<P: AsRef<Path>, Q: AsRef<Path>>(
+    snapshot_path: P,
+    bank: &Bank,
+    full_snapshot_slot: Slot,
+    snapshot_version: Option<SnapshotVersion>,
+    snapshot_package_output_path: Q,
+    archive_format: ArchiveFormat,
+    thread_pool: Option<&ThreadPool>,
+    maximum_snapshots_to_retain: usize,
+) -> Result<PathBuf>
+where
+    P: fmt::Debug,
+{
+    assert!(bank.is_complete());
+    assert!(bank.slot() > full_snapshot_slot);
+
+    bank.squash(); // Bank may not be a root
+    bank.force_flush_accounts_cache();
+
+    // bprumo TODO: Cleaning still needs to be figured out.  My understanding is that we cannot
+    // clean past the last full snapshot slot here, but more testing needs to be done.  For now,
+    // I'm not cleaning at all when creating an incremental snapshot, since a full snapshot must
+    // have been created first, which would have performed a clean then.  Below are the different
+    // `clean_accounts()` options:
+    //      bank.clean_accounts(true, false);
```

Comment:
> My understanding is that we cannot clean past the last full snapshot slot here, but more testing needs to be done.  

We've already cleaned up to the last full snapshot slot `F` when the full snapshot was taken no? When taking the incremental snapshot at `S > F` we have to make sure:

1) We don't clean any updates in slots `<= S` based on updates in roots `> S`

2) Any "cleanable" zero-lamport updates in root `Z`, where `F < Z < S` are not removed, where they typically might be (recall zero-lamport keys are cleanable when the refcount of that account is equal to the number of updates).  This is because if these zero-lamport accounts are missing from the AppendVecs, future incremental snapshots will be missing these updates. <\br>
I think this is the biggest divergence from current behavior where we might have cleaned up these zero lamport accounts when running clean up to slot `Z`, we can't do that anymore unless we introduce some other mechanism of tracking certain accounts have become zero (maybe some other structure for tracking zero lamport accounts needs to be added to the snapshot so that we can do this clean from the AppendVecs)?

Reply:
Yep yep. So I just pushed another commit that has incremental snapshots working, but relies on preventing cleaning past the last full snapshot slot. After chatting elsewhere, I put in a test to ensure we don't lose zero-lamport accounts that went to zero _after_ the last full snapshot slot. So that test is still good, but now I need to change the clean logic.

Instead of preventing cleaning past the last full snapshot slot, I'll be keeping track of all zero-lamport accounts that went to zero after the last full snapshot slot. So that when an Incremental Snapshot is taken, we'll always pull these back in, so there's no zombie resurrections in subsequent incremental snapshots.